extension/src/shared/context/WidgetContext.tsx

```diff
@@ -84,127 +71,149 @@ export const WidgetContext: FC<PropsWithChildren<Props>> = ({
   customSiteUrl,
   geojsonURL,
 }) => {
+  const [hideFeedbackState, setHideFeedbackState] = useHideFeedback();
   useEffect(() => {
-    if (!PLATEAU_API_URL && plateauUrl) {
-      setPlateauApiUrl(plateauUrl);
+    if (hideFeedback !== undefined && hideFeedback !== hideFeedbackState) {
+      setHideFeedbackState(hideFeedback);
     }
-  }, [plateauUrl]);
+  }, [hideFeedback, hideFeedbackState, setHideFeedbackState]);
 
+  const [plateauApiUrlState, setPlateauApiUrl] = usePlateauApiUrl();
   useEffect(() => {
-    if (!PROJECT_ID && projectId) {
-      setProjectId(projectId);
+    if (!plateauApiUrlState && plateauUrl) {
+      setPlateauApiUrl(plateauUrl);
     }
-  }, [projectId]);
+  }, [plateauUrl, plateauApiUrlState, setPlateauApiUrl]);
 
+  const [projectIdState, setProjectIdState] = useProjectId();
   useEffect(() => {
-    if (!GEO_API_URL && geoUrl) {
-      setGeoApiUrl(geoUrl);
+    if (!projectIdState && projectId) {
+      setProjectIdState(projectId);
     }
-  }, [geoUrl]);
+  }, [projectId, projectIdState, setProjectIdState]);
 
+  const [geoApiUrlState, setGeoApiUrlState] = useGeoApiUrl();
   useEffect(() => {
-    if (!GSI_TILE_URL && gsiTileURL) {
-      setGISTileURL(gsiTileURL);
+    if (!geoApiUrlState && geoUrl) {
+      setGeoApiUrlState(geoUrl);
     }
-  }, [gsiTileURL]);
+  }, [geoUrl, geoApiUrlState, setGeoApiUrlState]);
 
+  const [gsiTileURLState, setGSITileURLState] = useGsiTileUrl();
   useEffect(() => {
-    if (!GOOGLE_STREET_VIEW_API_KEY && googleStreetViewAPIKey) {
-      setGoogleStreetViewAPIKey(googleStreetViewAPIKey);
+    if (!gsiTileURLState && gsiTileURL) {
+      setGSITileURLState(gsiTileURL);
     }
-  }, [googleStreetViewAPIKey]);
+  }, [gsiTileURL, gsiTileURLState, setGSITileURLState]);
 
+  const [googleStreetViewAPIKeyState, setGoogleStreetViewAPIKeyState] = useGoogleStreetViewApiKey();
   useEffect(() => {
-    if (!geoClient && geoUrl) {
-      createGeoClient(geoUrl);
+    if (!googleStreetViewAPIKeyState && googleStreetViewAPIKey) {
+      setGoogleStreetViewAPIKeyState(googleStreetViewAPIKey);
     }
-  }, [geoUrl]);
+  }, [googleStreetViewAPIKey, googleStreetViewAPIKeyState, setGoogleStreetViewAPIKeyState]);
 
+  // optional (custom) state
+  const [cityNameState, setCityNameState] = useCityName();
   useEffect(() => {
-    if (HIDE_FEEDBACK === undefined && hideFeedback !== undefined) {
-      setHideFeedback(hideFeedback);
+    if (cityName && (!cityNameState || cityNameState !== cityName)) {
+      setCityNameState(cityName);
     }
-  }, [hideFeedback]);
+  }, [cityName, cityNameState, setCityNameState]);
 
+  const [cityCodeState, setCityCodeState] = useCityCode();
   useEffect(() => {
-    const url = inEditor ? catalogURLForAdmin || catalogUrl : catalogUrl;
-    if (url) {
-      createCatalogClient(url, inEditor ? plateauToken : undefined);
+    if (cityCode && (!cityCodeState || cityCodeState !== cityCode)) {
```

Comment:
I couldn't assign empty string. Should we check only undefined value?
```suggestion
    if (cityCode !== undefined && (!cityCodeState || cityCodeState !== cityCode)) {
```